server/asset/eth/common.go

```diff
@@ -54,12 +54,188 @@ const (
+	Encode() []byte
+}
+
+const (
+	// coin type id (2) + tx id (32) = 34
+	txCoinIDSize = 34
+	// coin type id (2) + address (20) + secret has (32) = 54
+	swapCoinIDSize = 54
+	// coin type id (2) + address (20) + amount (8) + nonce (8) = 38
+	amountCoinIDSize = 38
+)
+
+// TxCoinID identifies a coin by the transaction ID that was used to send it
+// to the smart contract. This type of ID is useful to identify coins that
+// were sent in transactions that have not yet been mined.
+type TxCoinID struct {
+	TxID [32]byte
+}
+
+// String creates a human readable string.
+func (c *TxCoinID) String() string {
+	return fmt.Sprintf("tx: %x", c.TxID)
+}
+
+// Encode creates a byte slice that can be decoded with DecodeCoinID.
+func (c *TxCoinID) Encode() []byte {
+	b := make([]byte, txCoinIDSize)
+	binary.BigEndian.PutUint16(b[:2], uint16(CIDTxID))
+	copy(b[2:], c.TxID[:])
+	return b
+}
+
+var _ CoinID = (*TxCoinID)(nil)
+
+// decodeTxCoinID decodes a byte slice into an TxCoinID struct.
+func decodeTxCoinID(coinID []byte) (*TxCoinID, error) {
+	if len(coinID) != txCoinIDSize {
+		return nil, fmt.Errorf("decodeTxCoinID: length expected %v, got %v",
+			txCoinIDSize, len(coinID))
+	}
+
+	flag := binary.BigEndian.Uint16(coinID[:2])
+	if CoinIDFlag(flag) != CIDTxID {
+		return nil, fmt.Errorf("decodeTxCoinID: flag expected %v, got %v",
+			flag, CIDTxID)
+	}
+
+	var txID [32]byte
+	copy(txID[:], coinID[2:])
+	return &TxCoinID{
+		TxID: txID,
+	}, nil
+}
+
+// SwapCoinID identifies a coin in a swap contract.
+type SwapCoinID struct {
+	ContractAddress common.Address
+	SecretHash      [32]byte
+}
+
+// String creates a human readable string.
+func (c *SwapCoinID) String() string {
+	return fmt.Sprintf("contract: %v, secret hash:%x",
+		c.ContractAddress, c.SecretHash)
+}
+
+// Encode creates a byte slice that can be decoded with DecodeCoinID.
+func (c *SwapCoinID) Encode() []byte {
+	b := make([]byte, swapCoinIDSize)
+	binary.BigEndian.PutUint16(b[:2], uint16(CIDSwap))
+	copy(b[2:22], c.ContractAddress[:])
+	copy(b[22:], c.SecretHash[:])
+	return b
+}
+
+var _ CoinID = (*SwapCoinID)(nil)
+
+// decodeSwapCoinID decodes a byte slice into an SwapCoinID struct.
+func decodeSwapCoinID(coinID []byte) (*SwapCoinID, error) {
+	if len(coinID) != swapCoinIDSize {
+		return nil, fmt.Errorf("decodeSwapCoinID: length expected %v, got %v",
+			txCoinIDSize, len(coinID))
+	}
+
+	flag := binary.BigEndian.Uint16(coinID[:2])
+	if CoinIDFlag(flag) != CIDSwap {
+		return nil, fmt.Errorf("decodeSwapCoinID: flag expected %v, got %v",
+			flag, CIDSwap)
+	}
+
+	var contractAddress [20]byte
+	var secretHash [32]byte
+	copy(contractAddress[:], coinID[2:22])
+	copy(secretHash[:], coinID[22:])
+	return &SwapCoinID{
+		ContractAddress: contractAddress,
+		SecretHash:      secretHash,
+	}, nil
+}
+
+// AmountCoinID is an identifier for a coin which has not yet been sent to the
+// swap contract.
+type AmountCoinID struct {
+	Address common.Address
+	Amount  uint64
+	Nonce   [8]byte
+}
```

Comment:
I think this type will also be used for server.FundingCoins since we are trying to portray? unspent value on the blockchain there as well, so glad it's here after all.

Reply:
Yeah that's why I wanted it to be accessible from both the server/client. The client sends these IDs to the server.